File: pkg-r/R/handoff_validation.R
validate_handoff_source <- function(source, handoff_type) {
  if (!S7::S7_inherits(handoff_type, HandoffType)) {
    cli::cli_abort("{.arg handoff_type} must be a <HandoffType> object.")
  }
  if (
    !is.character(source) ||
      length(source) != 1L ||
      is.na(source) ||
      !nzchar(trimws(source))
  ) {
    cli::cli_abort("Generated handoff source must be a non-empty string.")
  }
  if (identical(handoff_type@structure, "notebook-json")) {
    validate_notebook_handoff_source(source, handoff_type@language)
  }

  invisible(NULL)
}

validate_notebook_handoff_source <- function(source, language) {
  notebook <- tryCatch(
    jsonlite::parse_json(source),
    error = function(error) abort_invalid_notebook_source()
  )
  if (
    !is.list(notebook) ||
      is.null(names(notebook)) ||
      !is.list(notebook$cells) ||
      !is.list(notebook$metadata) ||
      !is.numeric(notebook$nbformat)
  ) {
    abort_invalid_notebook_source()
  }

  label <- handoff_language_label(language)
  kernelspec <- notebook$metadata$kernelspec
  actual <- if (is.list(kernelspec)) kernelspec$language else NULL
  if (!is.character(actual) || length(actual) != 1L || is.na(actual)) {
    cli::cli_abort("Generated notebook must declare a {label} kernelspec.")
  }
  if (tolower(actual) != tolower(language)) {
    cli::cli_abort(
      "Generated notebook must declare a {label} kernelspec, not {actual}."
    )
  }

  invisible(NULL)
}

abort_invalid_notebook_source <- function() {
  cli::cli_abort("Generated source is not valid notebook JSON.")
}

File: pkg-py/src/querychat/_handoff_validation.py
import nbformat
from nbformat.reader import NotJSONError

from ._handoff_types import LANGUAGES, HandoffType

HANDOFF_MIN_SOURCE_LENGTH = 200
HANDOFF_PROMPT_ECHO_PREFIX_LENGTH = 80


class HandoffValidationError(ValueError):
    """Generated handoff source violates its target contract."""


def validate_handoff_source(
    source: str,
    handoff_type: HandoffType,
    system_prompt: str,
) -> None:
    if not source.strip():
        raise HandoffValidationError("Generated handoff source is empty.")
    if handoff_type.structure == "notebook-json":
        validate_notebook_source(source, handoff_type)
    else:
        validate_handoff_substance(source, system_prompt)


def validate_handoff_substance(source: str, system_prompt: str) -> None:
    if len(source) < HANDOFF_MIN_SOURCE_LENGTH:
        raise HandoffValidationError(
            "Generated handoff source is too short to be a real handoff "
            f"({len(source)} characters)."
        )
    prompt_prefix = system_prompt.strip()[:HANDOFF_PROMPT_ECHO_PREFIX_LENGTH]
    if prompt_prefix and prompt_prefix in source:
        raise HandoffValidationError(
            "Generated handoff source echoes the system prompt instead of "
            "producing real content."
        )


def validate_notebook_source(
    source: str,
    handoff_type: HandoffType,
) -> None:
    try:
        notebook = nbformat.reads(source, as_version=4)
        nbformat.validate(notebook)
    except (NotJSONError, nbformat.ValidationError) as exc:
        raise HandoffValidationError(
            "Generated source is not valid notebook JSON."
        ) from exc

    kernelspec = notebook.metadata.get("kernelspec")
    actual = kernelspec.get("language") if kernelspec is not None else None
    expected = handoff_type.language
    if expected is None:
        raise HandoffValidationError(
            "Notebook validation requires a resolved R or Python language."
        )
    label = LANGUAGES[expected]
    if not isinstance(actual, str):
        raise HandoffValidationError(
            f"Generated notebook must declare a {label} kernelspec."
        )
    if actual.casefold() != expected.casefold():
        raise HandoffValidationError(
            f"Generated notebook must declare a {label} kernelspec, not {actual}."
        )
